### DISCUSSION TITLE
How can detect last retry?

### DISCUSSION THREAD

**Discussion by donghoon-khan** [spring-projects/spring-retry#362] [Q&A] [answered]
I am using Spring-retry-1.3.4, i want to detect the event before the last retry and insert logic.
I tried to use RetryListener, but I received an answer that it was impossible because RetryListener did not have retry policy.
Is there any other way?

### TestService.java
```java
@Retryable(
        value = {Exception.class},
        maxAttempts = 2
)
public void test1() {}

@Retryable(
        value = {Exception.class},
        maxAttempts = 3
)
public void test2() {}

@Retryable(
        value = {Exception.class},
        maxAttempts = 4
)
public void test3() {}
```

Like this, when I set maxAttempts differently using @Retryable annotation, I want to add a common logic before last attempt.

**Comment by artembilan** [comment]
See our original discussion: https://github.com/spring-projects/spring-retry/issues/361.

Not sure what else you'd like to hear from us that has driven you to still open this discussion...

**Comment by garyrussell** [comment, marked as answer]
There is no such mechanism available.

One work around would be to use a property for the max attempts, together with a label.

```java
@Value("${test3.max}") int test3Max;
...
@Retryable(
        value = {Exception.class},
        maxAttemptsExpression = "#{'${test3.max}'}",
        label = "test3"
)
public void test3() {}
```

...and use the same property in your listener.

The label is available in the `RetryContext` under the `RetryContext.NAME` property.

```java
	/**
	 * Retry context attribute name for reporting key. Can be used for reporting purposes,
	 * for instance in a retry listener, to accumulate data about the performance of a
	 * retry.
	 */
	String NAME = "context.name";
```

**Reply by garyrussell** [reply]
Or, even simpler, ...
```java
@Retryable(
        value = {Exception.class},
        maxAttempts= 4,
        label = "4"
)
```

Then parse the label in your listener to an int.

**Reply by e-ivaldi** [reply]
Hi @garyrussell, first, thanks for writing this down. Can I ask why have you decided to not provide some support for this use case ? I understand these workarounds exist, but they are still a workarounds (e.g. the label suggestion works fine but then you'd have two slightly different ways to get it working when using annotations vs using the programmatic approach, as in the latter one your listener would also have to know the the maximum number of attempts set in the policy)  and they still require to define both maxAttempts and the label . I understand that not all policies are limited by a number of max attempts but they could still offer a way pass that information to the retry context so that it would be available for clients to use.

Would something along the line of this [MR](https://github.com/e-ivaldi/spring-retry/pull/1/files) be terrible to get into the library itself ? Anyway - not a big deal as we have alternatives, just curious.

**Reply by garyrussell** [reply]
@e-ivaldi 

>...have you decided to not provide...

I have made no such decision; nobody has actually asked for the enhancement; this discussion does not constitute a feature request; it was simply a question, which I answered.

I have no objection to adding the feature; contributions, as always, are welcome.

**Reply by e-ivaldi** [reply]
Ah, okay. Thanks. I made the wrong assumption. In this case I'll check out the guidelines to see how can I contribute adding that as a feature.
